### PR TITLE
feat(compiler): implement style encapsulation for new view engine

### PR DESCRIPTION
**Pull request by tbosch**

### CODE REVIEW

File: modules/@angular/core/src/render/api.ts [hunk 1]
```diff
@@ -99,16 +92,52 @@ export abstract class Renderer {
...
+  abstract renderComponent(componentType: RenderComponentType): Renderer;
+}
+
+export interface ComponentRenderTypeV2 {
+  id: string;
+  encapsulation: ViewEncapsulation;
+  styles: (string|any[])[];
+  data: {[kind: string]: any[]};
+}
+;
```
Comment (tbosch) [inline modules/@angular/core/src/render/api.ts]: remove

File: modules/@angular/core/src/render/api.ts [hunk 2]
```diff
@@ -99,16 +92,52 @@ export abstract class Renderer {
...
+  abstract destroy(): void;
+  abstract createElement(name: string, namespace?: string): any;
+  abstract createComment(value: string): any;
+  abstract createText(value: string): any;
+  /**
+   * This property is allowed to be null / undefined,
+   * in which case the view engine won't call it.
+   * This is used as a performance optimization for production mode.
+   */
+  abstract destroyNode(node: any): void;
```
Comment (tbosch) [inline modules/@angular/core/src/render/api.ts]: change to `destroyNode: (node: an) => void | null`

File: modules/@angular/core/src/view/index.ts [hunk 3]
```diff
@@ -12,7 +12,7 @@ export {directiveDef, pipeDef, providerDef} from './provider';
 export {pureArrayDef, pureObjectDef, purePipeDef} from './pure_expression';
 export {queryDef} from './query';
 export {createComponentFactory} from './refs';
-export {initServicesIfNeeded} from './services';
+export {getCurrentDebugContext, initServicesIfNeeded} from './services';
```
Comment (tbosch) [inline modules/@angular/core/src/view/index.ts]: remove again

File: modules/@angular/core/src/view/provider.ts [hunk 4]
```diff
@@ -26,10 +28,20 @@ const InjectorRefTokenKey = tokenKey(Injector);
 
 const NOT_CREATED = new Object();
 
+let _renderCompCount = 0;
+
+function nextRenderCompId() {
```
Comment (tbosch) [inline modules/@angular/core/src/view/provider.ts]: wrong, id has to be passed to the `directiveDef` already!

File: modules/@angular/core/src/view/services.ts [hunk 5]
```diff
@@ -220,6 +234,15 @@ function camelCaseToDashCase(input: string): string {
   return input.replace(CAMEL_CASE_REGEXP, (...m: any[]) => '-' + m[1].toLowerCase());
 }
 
+function normalizeDebugBindingValue(value: any): string {
+  try {
+    // Limit the size of the value as otherwise the DOM just gets polluted.
+    return value ? value.toString().slice(0, 10) : value;
```
Comment (tbosch) [inline modules/@angular/core/src/view/services.ts]: maybe set to 20

File: modules/@angular/platform-server/src/server.ts [hunk 6]
```diff
@@ -42,8 +42,9 @@ export function _createConditionalRootRenderer(rootRenderer: any) {
   return isDevMode() ? new DebugDomRootRenderer(rootRenderer) : rootRenderer;
 }
 
-export function _createDebugRendererV2(renderer: RendererV2): RendererV2 {
-  return isDevMode() ? new DebugDomRendererV2(renderer) : renderer;
+export function _createDebugRendererV2(
```
Comment (tbosch) [inline modules/@angular/platform-server/src/server.ts]: remove

**Review by vicb — APPROVED** [review]

File: modules/@angular/core/src/render/api.ts (hunk 1 above)
Comment (tbosch) [inline modules/@angular/core/src/render/api.ts]: Done

File: modules/@angular/core/src/render/api.ts (hunk 2 above)
Comment (tbosch) [inline modules/@angular/core/src/render/api.ts]: Done

File: modules/@angular/core/src/view/index.ts (hunk 3 above)
Comment (tbosch) [inline modules/@angular/core/src/view/index.ts]: Done

File: modules/@angular/core/src/view/services.ts (hunk 5 above)
Comment (tbosch) [inline modules/@angular/core/src/view/services.ts]: Done

File: modules/@angular/platform-server/src/server.ts (hunk 6 above)
Comment (tbosch) [inline modules/@angular/platform-server/src/server.ts]: Done

File: modules/@angular/core/src/view/provider.ts (hunk 4 above)
Comment (tbosch) [inline modules/@angular/core/src/view/provider.ts]: Done

**Review by mhevery — APPROVED** [review]

**Pull request merged**